src/BackendRequest.cpp

```diff
@@ -99,9 +100,67 @@ namespace NextPVR
     }
     return false;
   }
-  std::vector<std::string> Request::Discovery()
+  bool Request::OneTimeSetup(void *hdl)
   {
-    std::vector<std::string> foundAddress;
+    // create user folder for channel icons and try and locate backend
+    #if defined(TARGET_WINDOWS)
+      #undef CreateDirectory
+    #endif
+    const std::string backend = "http://127.0.0.1:8866/service?method=recording.lastupdated|connection-timeout=2";
+    void* fileHandle = XBMC->OpenFile(backend.c_str(), READ_NO_CACHE);
+    if (fileHandle)
+    {
+      XBMC->CloseFile(fileHandle);
+      XBMC->CreateDirectory("special://userdata/addon_data/pvr.nextpvr/");
+      return true;
+    }
+    else
+    {
+      // couldn't find NextPVR on localhost so try and find on subnet
+      std::vector<std::vector<std::string>> foundAddress = Discovery();
+      int offset = 0;
+      if (!foundAddress.empty())
+      {
+        // found host using discovery protocol.
+        if (foundAddress.size() > 1)
+        {
+          // found multiple hosts let user choose
+          CHelper_libKODI_guilib* GUI = new CHelper_libKODI_guilib;
+          GUI->RegisterMe(hdl);
+          const char* entries[5];
+          int entry;
+          for (entry = 0; entry < 5 && entry < foundAddress.size(); entry++)
+          {
+            entries[entry] = foundAddress[entry][0].c_str();
+          }
+          offset = GUI->Dialog_Select(XBMC->GetLocalizedString(30187), entries, entry, 0);
+          delete GUI;
+          if (offset < 0)
+          {
+            XBMC->Log(LOG_INFO,"User canceled setup expect a failed install");
+            return false;
+          }
+        }
+        XBMC->CreateDirectory("special://userdata/addon_data/pvr.nextpvr/");
+        m_settings.UpdateServerPort(foundAddress[offset][0], std::stoi(foundAddress[offset][1]));
+        XBMC->QueueNotification(QUEUE_INFO, XBMC->GetLocalizedString(30182), m_settings.m_hostname.c_str(), m_settings.m_port);
+        char *settings = XBMC->TranslateSpecialProtocol("special://profile/addon_data/pvr.nextpvr/settings.xml");
+        // create settings.xml with host and port
+        void* settingsFile = XBMC->OpenFileForWrite(settings,false);
+        const char tmp[] = "<settings version=\"2\">\n</settings>\n";
+        XBMC->WriteFile(settingsFile,tmp, strlen(tmp));
+        XBMC->CloseFile(settingsFile);
+        m_settings.SaveSettings("host",m_settings.m_hostname);
+        m_settings.SaveSettings("port",std::to_string(m_settings.m_port));
+        XBMC->FreeString(settings);
```

Comment:
> 
> 
> Can you explain this, it seem really hacky and I don't think should be allowed.

I am creating the default settings.xml with discovered IP address and port.  If I don't do this when the user goes to edit the settings it will be 127.0.0.1 since the read will use the settings.xml default.   It is only created when the folder doesn't existing so Kodi hasn't had a chance to read it.  After this is loaded the API calls access the file ok.

I have been updating the settings.xml with the MAC address for WOL for over one year with now issue.

Reply:
What happens if a user resets settings to default?

Reply:
This is the user settings.xml so default="true" won't be set by settings.  The discovered IP is never the default 127.0.0.1 so it will act like a user entered value.  If the user does click defaults it will get default from the master.

Here is it in action https://www.youtube.com/watch?v=XTkQLdKGF4M

Reply:
It's still a hack though. Would it not be better to just discover it each time?

Reply:
Discoverying every time would be a hack and would confuse a user with 127.0.0.1 in settings when they want to customize turn on time shift etc.  It is also leads to potential issues if a different backend was discovered without proper updating. Finally discovery is slower then a socket open.

While you consider it a hack, updating the ADDON API to support configurable default settings for this is not a small undertaking.

Reply:
Agreed, it’s already in the C++ API, so the ability will come
When that does.

I don’t there there is any workaround for this currently outside of what you have done (I still don’t have to like it 😉).

Reply:
Ok I will keep this in and gladily migrate it when the new API comes out.